File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/recipes/domain/usecase/SaveRecipeUseCaseTest.java
package com.tarbadev.witchcraft.recipes.domain.usecase;

import com.tarbadev.witchcraft.recipes.domain.usecase.SaveRecipeUseCase;
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe;
import com.tarbadev.witchcraft.recipes.domain.repository.RecipeRepository;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class SaveRecipeUseCaseTest {
  private SaveRecipeUseCase saveRecipeUseCase;

  @Mock private RecipeRepository recipeRepository;

  @Before
  public void setUp() {
    saveRecipeUseCase = new SaveRecipeUseCase(recipeRepository);
    reset(recipeRepository);
  }

  @Test
  public void execute() {
    Recipe recipe = Recipe.builder().build();

    saveRecipeUseCase.execute(recipe);

    verify(recipeRepository).saveRecipe(recipe);
  }

  @Test
  public void execute_updatesRecipe() {
    Recipe recipe = Recipe.builder().id(123).build();

    saveRecipeUseCase.execute(recipe);

    verify(recipeRepository).updateRecipe(recipe);
  }
}

File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/recipes/domain/usecase/SaveRecipeUseCaseTest.kt
package com.tarbadev.witchcraft.recipes.domain.usecase

import com.nhaarman.mockitokotlin2.mock
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import com.tarbadev.witchcraft.recipes.domain.repository.RecipeRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.reset
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner

class SaveRecipeUseCaseTest {
    private lateinit var saveRecipeUseCase: SaveRecipeUseCase
    private val recipeRepository: RecipeRepository = mock()

    @BeforeEach
    fun setUp() {
        saveRecipeUseCase = SaveRecipeUseCase(recipeRepository)
        reset(recipeRepository)
    }

    @Test
    fun execute() {
        val recipe = Recipe()

        saveRecipeUseCase.execute(recipe)

        verify<RecipeRepository>(recipeRepository).saveRecipe(recipe)
    }

    @Test
    fun execute_updatesRecipe() {
        val recipe = Recipe(id = 123)

        saveRecipeUseCase.execute(recipe)

        verify<RecipeRepository>(recipeRepository).updateRecipe(recipe)
    }
}